lib/rate_map.h

```diff
@@ -23,13 +23,39 @@
 #include <stddef.h>
 #include <stdio.h>
 
+typedef struct {
+    double query_multiplier;
+    size_t num_lookups;
+    const double **lookups;
+} fast_search_t;
+
 typedef struct {
     size_t size;
     double *position;
     double *rate;
     double *cumulative_mass;
+    fast_search_t position_lookup;
 } rate_map_t;
 
+int fast_search_alloc(fast_search_t *self, const double *values, size_t n_values);
+int fast_search_free(fast_search_t *self);
+const double *fast_search_ptr_upper(fast_search_t *self, double query);
+const double *fast_search_ptr_strict_upper(fast_search_t *self, double query);
+
+inline size_t
+fast_search_idx_upper(fast_search_t *self, double query)
```

Comment:
Good question. Due to the compiler's freedom to move around what registers are used, I think it usually (depending on the caller) would result in fewer machine instructions. I'm hoping to do a quick investigation out of curiosity. Not that these few machine instructions are doing to make much diff.
And in my next commit I'll make the issue moot. The trade-off here doesn't need to be perf vs inlines in headers. All these ptr_* versions of the functions do not need to be available to multiple .c files. They only get used in one spot. So I've just moved this inline code to the .c files that use them. So it'll be just a hypothetical future trade-off in reduced code re-use.

Reply:
I was half-incorrect. There is still a perf trade-off avoiding inline functions in headers. One of the functions (fast_search_idx_strict_upper) that was inlined is used across .c files, namely rate_map.c and the unit tests.

I did an investigation and confirm that the answer to your question is: yes, it does make a difference, no, you do not get the same thing. Specifically, by not having `rate_map_position_to_mass` call `fast_search_idx_strict_upper` as an inline function, `rate_map_position_to_mass` is spending 79 instructions per call. In contrast, letting `fast_search_idx_strict_upper` be an inline function the result is 74 instructions per call. So 6% fewer instructions.

The reason for this is complicated and really interesting. I can write a longer answer with the details.

Anyway, not a big difference but it is there. The current commit has no inline functions in headers. But that does make it slightly less performant. An inline version of this with the 6% boost and inline function implementations at the end of headers files is here
https://github.com/castedo/msprime/compare/fast_search_lookup...castedo:inline_fast_search_lookup
in case you want to use that.